openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/strcompat/doc/man/man9/hpuxcompat.9.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) hpuxcompat.9.man,v 0.9.2.7 2006/04/05 09:43:32 brian Exp
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified 2006/04/05 09:43:32 by brian
.\"
.\" -------------------------------------------------------------------------
.\"
.\" hpuxcompat.9.man,v
.\" Revision 0.9.2.7  2006/04/05 09:43:32  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database strcompat.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so strcompat.macros
.\"
.\"
.TH HPUXCOMPAT 9 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ DDI/DKI"
.\"
.\"
.SH NAME
.B hpuxcompat
\- HP-UX \fISTREAMS\fR Compatibility
.\"
.\"
.SH SYNOPSIS
.PP
.B #define _HPUX_SOURCE
.br
.B #include <sys/ddi.h>
.\"
.\"
.SH OVERVIEW
.PP
.B Linux Fast-STREAMS
provides a compatibility module,
.BR streams-hpuxcompat.o ,
that provides compatibility functions that smooth porting of drivers from the
.I HP-UX 11i v2
.I STREAMS/UX
environment to
.BR "Linux Fast-STREAMS" .
The functions provided by this compatibility module are described here.
.\"
.\"
.SH DESCRIPTION
.PP
Many of the functions that are provided by
.IR HP-UX \(rg
are also provided by the core
.B streams.o
kernel module of
.BR "Linux Fast-STREAMS" .
Other
.IR HP-UX \(rg-specific
functions are provided by the
.B streams-hpuxcompat.o
compatibility kernel module.
.\"
.\"
.SS "Kernel Functions"
.PP
Kernel functions available in
.IR HP-UX \(rg
that are already available in
.B Linux
are as follows:
.HP
.BR bcopy (9)\ \-\ \c
copy byte strings
.PD 0
.HP
.BR bzero (9)\ \-\ \c
zero a byte string
.HP
.BR max (9)\ \-\ \c
determine the maximum of two integers
.HP
.BR min (9)\ \-\ \c
determine the minimum of two integers
.PD
.\"
.PP
Kernel functions available in
.IR HP-UX \(rg
that are available in
.B Linux Fast-STREAMS
(in the core
.I streams
kernel module) are as folllows:
.HP
.BR cmn_err (9)\ \-\ \c
print a kernel command error
.PD 0
.HP
.BR delay (9)\ \-\ \c
postpone the calling process for a number of clock ticks
.HP
.BR drv_getparm (9)\ \-\ \c
driver retrieve kernel parameter
.HP
.BR drv_priv (9)\ \-\ \c
check if the current process is priviledged
.HP
.BR getmajor (9)\ \-\ \c
get the internal major device number for a device.
.HP
.BR getminor (9)\ \-\ \c
get the extended minor device number for a device.
.HP
.BR kmem_alloc (9)\ \-\ \c
allocate kernel memory
.HP
.BR kmem_free (9)\ \-\ \c
deallocates kernel memory
.HP
.BR makedevice (9)\ \-\ \c
create a device from major and minor device numbers
.PD
.\"
.SS "\fISVR 4.2 MP\fP Kernel Functions"
.PP
Kernel functions available in
.IR HP-UX \(rg
that are available in
.B Linux Fast-STREAMS
(in the
.I "SVR 4.2 MP"
compatibility kernel module,
.IR streams-svr4compat )
are as folllows:
.HP
.BR itimeout (9)\ \-\ \c
start a timer at a specified priority level
.PD 0
.HP
.BR LOCK (9)\ \-\ \c
acquire a basic lock
.HP
.BR LOCK_ALLOC (9)\ \-\ \c
allocate and initialize a basic lock
.HP
.BR LOCK_DEALLOC (9)\ \-\ \c
deallocate a dynamically allocated basic lock
.HP
.BR major (9)\ \-\ \c
get the major device number for a device
.HP
.BR makedev (9)\ \-\ \c
make a device number from a major and minor device number
.HP
.BR minor (9)\ \-\ \c
get the minor device number for a device
.HP
.BR sleep (9)\ \-\ \c
put a process to sleep
.HP
.BR spl0 (9)\ \-\ \c
set priority level
.HP
.BR spl1 (9)\ \-\ \c
set priority level
.HP
.BR spl2 (9)\ \-\ \c
set priority level
.HP
.BR spl3 (9)\ \-\ \c
set priority level
.HP
.BR spl4 (9)\ \-\ \c
set priority level
.HP
.BR spl5 (9)\ \-\ \c
set priority level
.HP
.BR spl6 (9)\ \-\ \c
set priority level
.HP
.BR spl7 (9)\ \-\ \c
set priority level
.HP
.BR spl (9)\ \-\ \c
set priority level
.HP
.BR splx (9)\ \-\ \c
set priority level
.HP
.BR SV_ALLOC (9)\ \-\ \c
allocate and initialize a synchronization variable
.HP
.BR SV_BROADCAST (9)\ \-\ \c
wake one or all processes sleeping on a synchronization variable
.HP
.BR SV_DEALLOC (9)\ \-\ \c
deallocate a synchronization variable
.HP
.BR SV_WAIT (9)\ \-\ \c
wait on a sychronization variable or signal
.HP
.BR SV_WAIT_SIG (9)\ \-\ \c
wait on a sychronization variable or signal
.HP
.BR TRYLOCK (9)\ \-\ \c
attempt to acquire a basic lock
.HP
.BR UNLOCK (9)\ \-\ \c
release a basic lock
.HP
.BR vtop (9)\ \-\ \c
.HP
.BR wakeup (9)\ \-\ \c
.PD
.\"
.SS "Core \fISTREAMS\fP Functions"
.PP
.I STREAMS
functions that are already available in
.B Linux Fast STREAMS
(in the core
.I streams
kernel module) are as follows:
.HP
.BR adjmsg (9)\ \-\ \c
trim bytes from the front or back of a \fISTREAMS\fP message
.PD 0
.HP
.BR allocb (9)\ \-\ \c
allocate a \fISTREAMS\fP message and data block
.HP
.BR backq (9)\ \-\ \c
find the upstream or downstream queue
.HP
.BR bcanput (9)\ \-\ \c
test flow control on a \fISTREAMS\fP message queue
.HP
.BR bcanputnext (9)\ \-\ \c
test flow control on a \fISTREAMS\fP message queue
.HP
.BR bufcall (9)\ \-\ \c
install a buffer callback
.HP
.BR canput (9)\ \-\ \c
test flow control on a \fISTREAMS\fP message queue
.HP
.BR canputnext (9)\ \-\ \c
test flow control on a \fISTREAMS\fP message queue
.HP
.BR copyb (9)\ \-\ \c
copy a \fISTREAMS\fP message block
.HP
.BR copymsg (9)\ \-\ \c
copy a \fISTREAMS\fP message
.HP
.BR datamsg (9)\ \-\ \c
tests a \fISTREAMS\fP message type for data
.HP
.BR dupb (9)\ \-\ \c
duplicate a \fISTREAMS\fP message block
.HP
.BR dupmsg (9)\ \-\ \c
duplicate a \fISTREAMS\fP message
.HP
.BR enableok (9)\ \-\ \c
allows a \fISTREAMS\fP message queue to be scheduled
.HP
.BR esballoc (9)\ \-\ \c
allocate a \fISTREAMS\fP message and data block with a caller supplied data buffer
.HP
.BR esbbcall (9)\ \-\ \c
install a buffer callback for an extended \fISTREAMS\fR message block
.HP
.BR flushband (9)\ \-\ \c
flushes band \fISTREAMS\fP messages from a message queue
.HP
.BR flushq (9)\ \-\ \c
flushes messages from a \fISTREAMS\fP message queue
.HP
.BR freeb (9)\ \-\ \c
frees a \fISTREAMS\fP message block
.HP
.BR freemsg (9)\ \-\ \c
frees a \fISTREAMS\fP message
.HP
.BR freezestr (9)\ \-\ \c
freeze the state of a stream queue
.HP
.BR getadmin (9)\ \-\ \c
get the administration function pointer for a \fISTREAMS\fP module
.HP
.BR getmid (9)\ \-\ \c
get the \fISTREAMS\fP module id for a name
.HP
.BR getq (9)\ \-\ \c
gets a message from a \fISTREAMS\fP message queue
.HP
.BR insq (9)\ \-\ \c
inserts a message into a \fISTREAMS\fP message queue
.HP
.BR linkb (9)\ \-\ \c
link a message block to a \fISTREAMS\fP message
.HP
.BR msgdsize (9)\ \-\ \c
calculate the size of the data in a \fISTREAMS\fP message
.HP
.BR msgppullup (9)\ \-\ \c
pull up bytes in a \fISTREAMS\fR message
.HP
.BR noenable (9)\ \-\ \c
disables a \fISTREAMS\fP message queue from being scheduled
.HP
.BR OTHERQ (9)\ \-\ \c
return the other queue of a \fISTREAMS\fP queue pair
.HP
.BR pcmsg (9)\ \-\ \c
test a data block message type for priority control
.HP
.BR pullupmsg (9)\ \-\ \c
pull up the bytes in a \fISTREAMS\fP message
.HP
.BR put (9)\ \-\ \c
invoke the put procedure for a \fISTREAMS\fP module or driver with a \fISTREAMS\fP message
.HP
.BR putbq (9)\ \-\ \c
put a message back on a \fISTREAMS\fP message queue
.HP
.BR putctl (9)\ \-\ \c
put a control message on a \fISTREAMS\fP message queue
.HP
.BR putctl1 (9)\ \-\ \c
put a 1 byte control message on a \fISTREAMS\fP message queue
.HP
.BR putctl2 (9)\ \-\ \c
put a two byte control message on a \fISTREAMS\fR message queue
.HP
.BR putnext (9)\ \-\ \c
put a message on the downstream \fISTREAMS\fP message queue
.HP
.BR putnextctl (9)\ \-\ \c
put a control message on the downstream \fISTREAMS\fR message queue
.HP
.BR putnextctl1 (9)\ \-\ \c
put a 1 byte control message on the downstream \fISTREAMS\fR message queue
.HP
.BR putnextctl2 (9)\ \-\ \c
put a two byte control message on the downstream \fISTREAMS\fR message queue
.HP
.BR putq (9)\ \-\ \c
put a message on a \fISTREAMS\fP message queue
.HP
.BR qenable (9)\ \-\ \c
schedules a \fISTREAMS\fP message queue service routine
.HP
.BR qprocsoff (9)\ \-\ \c
disables \fISTREAMS\fR message queue processing for multi-processing
.HP
.BR qprocon (9)\ \-\ \c
enables a \fISTREAMS\fR message queue for multi-processing
.HP
.BR qreply (9)\ \-\ \c
replys to a message from a \fISTREAMS\fP message queue
.HP
.BR qsize (9)\ \-\ \c
return the number of messages on a queue
.HP
.BR RD (9)\ \-\ \c
return the read queue of a \fISTREAMS\fP queue pair
.HP
.BR rmvb (9)\ \-\ \c
remove a message block from a \fISTREAMS\fP message
.HP
.BR rmvq (9)\ \-\ \c
remove a message from a \fISTREAMS\fP message queue
.HP
.BR SAMESTR (9)\ \-\ \c
test for \fISTREAMS\fR pipe or fifo
.HP
.BR splstr (9)\ \-\ \c
set priority levels
.HP
.BR strlog (9)\ \-\ \c
pass a message to the \fISTREAMS\fP logger
.HP
.BR strqget (9)\ \-\ \c
gets information about a \fISTREAMS\fP message queue
.HP
.BR strqset (9)\ \-\ \c
sets attributes of a \fISTREAMS\fR message queue
.HP
.BR testb (9)\ \-\ \c
test if a \fISTREAMS\fP message can be allocated
.HP
.BR timeout (9)\ \-\ \c
start a timer
.HP
.BR unbufcall (9)\ \-\ \c
removes a \fISTREAMS\fP buffer callback
.HP
.BR unfreezestr (9)\ \-\ \c
thaw the state of a stream queue
.HP
.BR unlinkb (9)\ \-\ \c
unlink a message block from a \fISTREAMS\fP message
.HP
.BR untimeout (9)\ \-\ \c
stop a timer
.HP
.BR unweldq  (9)\ \-\ \c
unweld two queues
.HP
.BR weldq  (9)\ \-\ \c
weld two (or four) queues together
.HP
.BR WR (9)\ \-\ \c
return the write queue of a \fISTREAMS\fR queue pair
.PD
.\"
.SS "Specific \fISTREAMS\fP Functions"
.PP
Functions that are available in the
.B Linux Fast-STREAMS
.IR HP-UX \(rg
compatibility module,
.BR streams-hpuxcompat.o ,
are as follows:
.HP
.BR get_sleep_lock (9)\ \-\ \c
provide access to the global sleep lock
.PD 0
.HP
.BR streams_get_sleep_lock (9)\ \-\ \c
provide access to the global sleep lock
.HP
.BR streams_mpsleep (9)\ \-\ \c
put a process to sleep
.HP
.BR streams_put (9)\ \-\ \c
invoke the put procedure for a \fISTREAMS\fP module or driver with a \fISTREAMS\fP message
.PD
.\"
.\".SS "\fISTREAMS\fP Configuration and Registration Functions"
.PP
Module and driver configuration functions that are available in the
.B Linux Fast-STREAMS
.IR HP-UX \(rg
compatibility module,
.IR streams-hpuxcompat ,
are as follows:
.HP
.BR str_install (9)\ \-\ \c
install or uninstall a \fISTREAMS\fP module or driver
.PD 0
.HP
.BR str_install_HPUX (9)\ \-\ \c
install or uninstall a \fISTREAMS\fP module or driver
.PD
.\"
.\"
.SH USAGE
.PP
For usage an example on the various functions, see the individual manual pages.
.\"
.\"
.SH CONTEXT
.PP
These functions can be executed from various contexts as described in each separate manual page.
.\"
.\"
.SH MP-STREAMS
.PP
Most of these functions are MP-safe; however,
.IR HP-UX \(rg
only provides stubs for some of the
.I SVR 4.2 MP
functions, such as:
.HP
.BR qprocson (9)\ \-\ \c
enables a \fISTREAMS\fR message queue for multi-processing
.PD 0
.HP
.BR qprocsoff (9)\ \-\ \c
disables \fISTREAMS\fR message queue processing for multi-processing
.HP
.BR freezestr (9)\ \-\ \c
freeze the state of a stream queue
.HP
.BR unfreezestr (9)\ \-\ \c
thaw the state of a stream queue
.PD
.PP
.IR HP-UX \(rg
expects that the apporpriate level of synchronization be used instead.
Because of this, it is recommended that
.IR HP-UX \(rg
modules and drivers port directly to the
.B Linux Fast-STREAMS
environment, not be run at synchronization level
.BR SQLVL_NOP .
See
.BR str_install_HPUX (9).
.\"
.\"
.SH NOTICES
.PP
.B _HPUX_SOURCE
must be defined before including DDI/DKI include files to expose
.IR HP-UX \(rg-specific
symbols.
To use
.IR HP-UX \(rg-specific
functions,
.B Linux Fast-STREAMS
must also have been compiled with
.B CONFIG_STREAMS_COMPAT_HPUX
or
.B CONFIG_STREAMS_COMPAT_HPUX_MODULE
defined.
When compiled with
.B CONFIG_STREAMS_COMPAT_HPUX
defined,
.IR HP-UX \(rg
compatibility will become part of the
.B Linux Fast-STREAMS
.B streams.o
kernel module.
When compiled with
.B CONFIG_STREAMS_COMPAT_HPUX_MODULE
defined,
.IR HP-UX \(rg
compatibility will be placed in a separate
.B streams-hpuxcompat.o
kernel module that will demand load when any
.I STREAMS
driver or module that uses the
.IR HP-UX \(rg-specific
functions loads.
.\"
.\"
.SH IMPLEMENTATION
.PP
.PP
Core functions are implemented within the
.B Linux Fast-STREAMS
framework.
.IR AIX \(rg-specific
functions are implemented as described in the separate manual pages describing each facility.
.\"
.\"
.SH "SEE ALSO"
.PP
Core functions:
.BR adjmsg (9),
.BR allocb (9),
.BR backq (9),
.BR bcanput (9),
.BR bcanputnext (9),
.BR bcopy (9),
.BR bufcall (9),
.BR bzero (9),
.BR canput (9),
.BR canputnext (9),
.BR cmn_err (9),
.BR copyb (9),
.BR copymsg (9),
.BR datamsg (9),
.BR delay (9),
.BR drv_getparm (9),
.BR drv_priv (9),
.BR dupb (9),
.BR dupmsg (9),
.BR enableok (9),
.BR esballoc (9),
.BR esbbcall (9),
.BR flushband (9),
.BR flushq (9),
.BR freeb (9),
.BR freemsg (9),
.BR freezestr (9),
.BR getadmin (9),
.BR getmid (9),
.BR getmajor (9),
.BR getminor (9),
.BR getq (9),
.BR insq (9),
.BR itimeout (9),
.BR kmem_alloc (9),
.BR kmem_free (9),
.BR linkb (9),
.BR LOCK (9),
.BR LOCK_ALLOC (9),
.BR LOCK_DEALLOC (9),
.BR major (9),
.BR makedev (9),
.BR makedevice (9),
.BR max (9),
.BR min (9),
.BR minor (9),
.BR msgdsize (9),
.BR msgppullup (9),
.BR noenable (9),
.BR OTHERQ (9),
.BR pcmsg (9),
.BR pullupmsg (9),
.BR put (9),
.BR putbq (9),
.BR putctl (9),
.BR putctl1 (9),
.BR putctl2 (9),
.BR putnext (9),
.BR putnextctl (9),
.BR putnextctl1 (9),
.BR putnextctl2 (9),
.BR putq (9),
.BR qenable (9),
.BR qprocsoff (9),
.BR qprocon (9),
.BR qreply (9),
.BR qsize (9),
.BR RD (9),
.BR rmvb (9),
.BR rmvq (9),
.BR SAMESTR (9),
.BR sleep (9),
.BR spln (9),
.BR splstr (9),
.BR strlog (9),
.BR strqget (9),
.BR strqset (9),
.BR SV_ALLOC (9),
.BR SV_BROADCAST (9),
.BR SV_DEALLOC (9),
.BR SV_WAIT (9),
.BR SV_WAIT_SIG (9),
.BR testb (9),
.BR timeout (9),
.BR TRYLOCK (9),
.BR unbufcall (9),
.BR unfreezestr (9),
.BR unlinkb (9),
.BR UNLOCK (9),
.BR untimeout (9),
.BR unweldq  (9),
.BR vtop (9),
.BR wakeup (9),
.BR weldq  (9),
.BR WR (9).
.PP
Compatibility functions:
.BR get_sleep_lock (9),
.BR streams_get_sleep_lock (9),
.BR streams_mpsleep (9),
.BR streams_put (9),
.BR str_install_HPUX (9).
.PP
General:
.BR STREAMS (9).
.\"
.\"
.SH BUGS
.PP
The
.IR HP-UX \(rg
compatibility suite has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
These functions are compatible with
.IR HP-UX \(rg
with the following portability considerations:
.IP \(em \w'\(em\(em'u
.IR HP-UX \(rg-specific
functions should not be used by portable
.I STREAMS
drivers and modules.
The more common
.I SVR 4.2 MP
functions should be used in their stead.
.IP \(em
.IR HP-UX \(rg-specific
functions are provided for source level compatibility with
.IR HP-UX \(rg
for the purpose of porting

openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/strcompat/src/kernel/hpuxcompat.c
/*****************************************************************************

 @(#) hpuxcompat.c,v OpenSS7-0_9_2_D_rc2(0.9.2.22) 2006/07/07 20:49:55

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2005  OpenSS7 Corporation <http://www.openss7.com>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; either version 2 of the License, or (at your option) any later
 version.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 675 Mass
 Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified 2006/07/07 20:49:55 by brian

 *****************************************************************************/

#ident "@(#) hpuxcompat.c,v OpenSS7-0_9_2_D_rc2(0.9.2.22) 2006/07/07 20:49:55"

static char const ident[] =
    "hpuxcompat.c,v OpenSS7-0_9_2_D_rc2(0.9.2.22) 2006/07/07 20:49:55";

/* 
 *  This is my solution for those who don't want to inline GPL'ed functions or
 *  who don't use optimizations when compiling or specifies
 *  -fnoinline-functions or something of the like.  This file implements all
 *  of the extern inlines from the header files by just including the header
 *  files with the functions declared 'inline' instead of 'extern inline'.
 *
 *  There are implemented here in a separate object, out of the way of the
 *  modules that don't use them.
 */

#define __HPUX_EXTERN_INLINE __inline__ streamscall

#define _HPUX_SOURCE

#include "sys/os7/compat.h"

#define HPUXCOMP_DESCRIP	"UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define HPUXCOMP_COPYRIGHT	"Copyright (c) 1997-2005 OpenSS7 Corporation.  All Rights Reserved."
#define HPUXCOMP_REVISION	"LfS hpuxcompat.c,v OpenSS7-0_9_2_D_rc2(0.9.2.22) 2006/07/07 20:49:55"
#define HPUXCOMP_DEVICE		"HP-UX 11i v2 Compatibility"
#define HPUXCOMP_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define HPUXCOMP_LICENSE	"GPL"
#define HPUXCOMP_BANNER		HPUXCOMP_DESCRIP	"\n" \
				HPUXCOMP_COPYRIGHT	"\n" \
				HPUXCOMP_REVISION	"\n" \
				HPUXCOMP_DEVICE		"\n" \
				HPUXCOMP_CONTACT	"\n"
#define HPUXCOMP_SPLASH		HPUXCOMP_DEVICE		" - " \
				HPUXCOMP_REVISION	"\n"

#ifdef CONFIG_STREAMS_COMPAT_HPUX_MODULE
MODULE_AUTHOR(HPUXCOMP_CONTACT);
MODULE_DESCRIPTION(HPUXCOMP_DESCRIP);
MODULE_SUPPORTED_DEVICE(HPUXCOMP_DEVICE);
MODULE_LICENSE(HPUXCOMP_LICENSE);
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-hpuxcompat");
#endif
#endif

static lock_t sleep_lock = SPIN_LOCK_UNLOCKED;

/**
 *  streams_get_sleep_lock: - acquire the global sleep lock
 *  @event:	the event which will be later passed to sleep
 *
 *  streams_get_sleep_lock() provides access to a global spinlock_t that may be used
 *  by all threads entering a wait queue to avoid race conditions between
 *  threads entering the wait queue.
 *
 *  Return Value:streams_get_sleep_lock() returns a pointer to the global sleep lock.
 */
lock_t *
streams_get_sleep_lock(caddr_t event)
{
	(void) event;
	return &sleep_lock;
}

EXPORT_SYMBOL_NOVERS(streams_get_sleep_lock);	/* hpux/ddi.h */

lock_t *get_sleep_lock(caddr_t event) __attribute__ ((alias("streams_get_sleep_lock")));

EXPORT_SYMBOL_NOVERS(get_sleep_lock);	/* hpux/ddi.h */

#ifdef LFS
/**
 *  streams_put: - deferred call to a STREAMS module qi_putp() procedure.
 *  @func:  put function (often the put() function)
 *  @q:	    queue against which to defer the call
 *  @mp:    message block to pass to the callback function
 *  @priv:  private data to pass to the callback function (often @q)
 *
 *  streams_put() will defer the function @func until it can synchronize with @q.  Once the @q has
 *  been syncrhonized, the STREAMS scheduler will call the callback function @func with arguments
 *  @priv and @mp.  streams_put() is closely related to qwrite() below.
 *
 *  Notices: @func will be called by the STREAMS executive on the same CPU as the CPU that called
 *  streams_put().  @func is guarateed not to run until the caller exits or preempts.
 *
 *  Usage: streams_put() is intended to be called from contexts outside of the STREAMS scheduler
 *  (e.g. interrupt service routines) where @func is intended to run under the STREAMS scheduler.
 *
 *  Examples: streams_put((void *)&put, q, mp, q) will effect the put() STREAMS utility, but always
 *  guaranteed to be executed within the STREAMS scheduler.
 */
void
streams_put(streams_put_t func, queue_t *q, mblk_t *mp, void *priv)
{
	extern void __strfunc(void (*func)(void *, mblk_t *), queue_t *q, mblk_t *mp, void *arg);
	__strfunc(func, q, mp, priv);
}

EXPORT_SYMBOL_NOVERS(streams_put);	/* hpux/ddi.h */
#endif

int
str_install_HPUX(struct stream_inst *inst)
{
#ifdef LIS
	switch (inst->inst_flags & STR_TYPE_MASK) {
	case STR_IS_DEVICE:
	{
		int err;

		if (inst->inst_major == -1)
			inst->inst_major = 0;
		if ((err = lis_register_strdev(inst->inst_major, &inst->inst_str_tab, 255,
					       inst->name)) > 0)
			inst->inst_major = err;
		return (err < 0 ? -err : 0);
	}
	case STR_IS_MODULE:
	{
		int err;

		if ((err = lis_register_strmod(&inst->inst_str_tab, inst->name)) > 0)
			inst->inst_major = err;
		return (err < 0 ? -err : 0);
	}
	default:
		return (EINVAL);
	}
#endif
#ifdef LFS
	switch (inst->inst_flags & STR_TYPE_MASK) {
	case STR_IS_DEVICE:
	{
		struct cdevsw *cdev;
		int err;

		if (!(cdev = kmem_zalloc(sizeof(*cdev), KM_NOSLEEP)))
			return (ENOMEM);
		cdev->d_name = inst->name;
		cdev->d_str = &inst->inst_str_tab;
		/* build flags */
		cdev->d_flag = 0;
		if (inst->inst_flags & STR_QSAFETY) {
			cdev->d_flag |= D_SAFE;
		}
		switch ((cdev->d_sqlvl = inst->inst_sync_level)) {
		case SQLVL_NOP:
			cdev->d_flag |= D_MP;
			break;
		case SQLVL_QUEUE:
			cdev->d_flag |= D_MTPERQ;
			break;
		case SQLVL_QUEUEPAIR:
			cdev->d_flag |= D_MTQPAIR;
			break;
		case SQLVL_MODULE:
			cdev->d_flag |= D_MTPERMOD;
			break;
		case SQLVL_ELSEWHERE:
			cdev->d_flag |= D_MTOUTPERIM;
			break;
		case SQLVL_GLOBAL:
			/* can't really support this, but its only used for debug anyway */
			cdev->d_flag &= ~D_MP;
			break;
		case SQLVL_DEFAULT:
			cdev->d_flag |= D_MTPERMOD;
			break;
		}
		if ((err = register_strdev(cdev, inst->inst_major)) < 0) {
			kmem_free(cdev, sizeof(*cdev));
			return (-err);
		}
		inst->inst_major = err;
		return (0);
	}
	case STR_IS_MODULE:
	{
		struct fmodsw *fmod;
		int err;

		if (!(fmod = kmem_zalloc(sizeof(*fmod), KM_NOSLEEP)))
			return (ENOMEM);
		fmod->f_name = inst->name;
		fmod->f_str = &inst->inst_str_tab;
		/* build flags */
		fmod->f_flag = 0;
		if (inst->inst_flags & STR_QSAFETY) {
			fmod->f_flag |= D_SAFE;
		}
		switch ((fmod->f_sqlvl = inst->inst_sync_level)) {
		case SQLVL_NOP:
			fmod->f_flag |= D_MP;
			break;
		case SQLVL_QUEUE:
			fmod->f_flag |= D_MTPERQ;
			break;
		case SQLVL_QUEUEPAIR:
			fmod->f_flag |= D_MTQPAIR;
			break;
		case SQLVL_MODULE:
			fmod->f_flag |= D_MTPERMOD;
			break;
		case SQLVL_ELSEWHERE:
			fmod->f_flag |= D_MTOUTPERIM;
			break;
		case SQLVL_GLOBAL:
			/* can't really support this, but its only used for debug anyway */
			fmod->f_flag &= ~D_MP;
			break;
		case SQLVL_DEFAULT:
			fmod->f_flag |= D_MTPERMOD;
			break;
		}
		if ((err = register_strmod(fmod)) < 0) {
			kmem_free(fmod, sizeof(*fmod));
			return (-err);
		}
		inst->inst_major = err;
		return (0);
	}
	default:
		return (EINVAL);
	}
#endif
}

EXPORT_SYMBOL_NOVERS(str_install_HPUX);

int
str_uninstall(struct stream_inst *inst)
{
#ifdef LIS
	switch (inst->inst_flags & STR_TYPE_MASK) {
	case STR_IS_DEVICE:
		return lis_unregister_strdev(inst->inst_major);
	case STR_IS_MODULE:
		return lis_unregister_strmod(&inst->inst_str_tab);
	default:
		return (EINVAL);
	}
#endif
#ifdef LFS
	switch (inst->inst_flags & STR_TYPE_MASK) {
	case STR_IS_DEVICE:
	{
		struct cdevsw *cdev;
		int err;

		if ((cdev = cdev_str(&inst->inst_str_tab)) == NULL)
			return (ENOENT);
		if ((err = unregister_strdev(cdev, inst->inst_major)) == 0)
			kmem_free(cdev, sizeof(*cdev));
		return (-err);
	}
	case STR_IS_MODULE:
	{
		struct fmodsw *fmod;
		int err;

		if ((fmod = fmod_str(&inst->inst_str_tab)) == NULL)
			return (ENOENT);
		if ((err = unregister_strmod(fmod)) == 0)
			kmem_free(fmod, sizeof(fmod));
		return (-err);
	}
	default:
		return (EINVAL);
	}
#endif
}

EXPORT_SYMBOL_NOVERS(str_uninstall);

#ifdef CONFIG_STREAMS_COMPAT_HPUX_MODULE
static
#endif
int __init
hpuxcomp_init(void)
{
#ifdef CONFIG_STREAMS_COMPAT_HPUX_MODULE
	printk(KERN_INFO HPUXCOMP_BANNER);
#else
	printk(KERN_INFO HPUXCOMP_SPLASH);
#endif
	return (0);
}

#ifdef CONFIG_STREAMS_COMPAT_HPUX_MODULE
static
#endif
void __exit
hpuxcomp_exit(void)
{
	return;
}

#ifdef CONFIG_STREAMS_COMPAT_HPUX_MODULE
module_init(hpuxcomp_init);
module_exit(hpuxcomp_exit);
#endif
